juju/model.py

```diff
@@ -3212,6 +3245,222 @@ def _raise_for_status(entities: dict[str, list[str]], status: Any):
+
+        for unit_name, unit in units.items():
+            assert unit.workload_status
+            if unit.workload_status.status == "error" and raise_on_error:
+                raise JujuUnitError(
+                    f"{unit_name!r} workload has errored: {unit.workload_status.info!r}"
+                )
+
+        for app_name in apps:
+            app = full_status.applications[app_name]
+            assert isinstance(app, ApplicationStatus)
+            assert app.status
+            if app.status.status == "error" and raise_on_error:
+                raise JujuAppError(f"{app_name!r} has errored: {app.status.info!r}")
+
+        for unit_name, unit in units.items():
+            assert unit.workload_status
+            if unit.workload_status.status == "blocked" and raise_on_blocked:
+                raise JujuUnitError(
+                    f"{unit_name!r} workload is blocked: {unit.workload_status.info!r}"
+                )
+
+        for app_name in apps:
+            app = full_status.applications[app_name]
+            assert isinstance(app, ApplicationStatus)
+            assert app.status
+            if app.status.status == "blocked" and raise_on_blocked:
+                raise JujuAppError(f"{app_name!r} is blocked: {app.status.info!r}")
+
+        for unit_name, unit in units.items():
+            assert unit.agent_status
+            idle_times.setdefault(unit_name, now)
+            if unit.agent_status.status != "idle":
+                idle_times[unit_name] = now
+
+        for app_name in apps:
+            ready_units = []
+            app = full_status.applications[app_name]
+            assert isinstance(app, ApplicationStatus)
+            for unit in _app_units(full_status, app_name).values():
+                assert unit.agent_status
+                assert unit.workload_status
+
+                if unit.agent_status.status != "idle":
+                    continue
+                if status and unit.workload_status.status != status:
+                    continue
+
+                ready_units.append(unit)
+
+            if wait_for_exact_units is None and len(ready_units) < _wait_for_units:
+                logger.info(
+                    "Waiting for app %r units %s/%s",
+                    app_name,
+                    len(ready_units),
+                    _wait_for_units,
+                )
+                return False
+
+            if (
+                wait_for_exact_units is not None
+                and len(ready_units) != wait_for_exact_units
+            ):
+                logger.info(
+                    "Waiting for app %r units %s/%s",
+                    app_name,
+                    len(ready_units),
+                    _wait_for_units,
+                )
+                return False
+
+        if busy := [n for n, t in idle_times.items() if expected_idle_since < t]:
+            logger.info("Waiting for %s to be idle enough", busy)
+            return False
+
+        return True
```

Comment:
Below I've pasted my understanding of the new and original wait for idle logic. Differences in order of operations are:
- original could raise errors before all apps are available, new can't
- original could return early once enough units were ready (assuming no errors had been accumulated in that pass through the apps) -- new will check for all errors before checking if units are ready (an improvement I think)

I suspect that both of these divergences are fine.

---

new

0. do all apps exist? if not, stop and request to check again later
1. check all apps and units for errors in the appropriate order, raising immediately if the user cares
2. for each app, figure out which units are ready -- if there are less than we're waiting for, stop and request to check again later
3. check if units have been idle for long enough -- if not, stop and request to heck again later
4. we're done

original

0. iterate over each app
1. does the app exist yet? if not, stop processing this app and request to iterate again
2. is the app in error/blocked and is that a problem? record it if so
3. are we waiting for enough units, and if so are there enough?
    - if there aren't, stop processing this app
    - if there are, stop processing apps entirely (but we'll start again if any earlier apps requested it -- not any later ones though -- smells like a bug)
4. now check the app's units for errors/blocked and record it if the user cares
5. now figure out if the app status means we need to start again
6. after stopping iteration
    - if any errors were recorded, raise the highest priority one
    - if any apps needed to start again, restart iteration, otherwise we're done

Reply:
I'm refactoring this mess... wait for it!

Reply:
Please have a look now :)

Reply:
your reading of the old loop is pretty close, but I think I've found one caveat: https://matrix.to/#/!UPTldjWgQImSaCAjYf:ubuntu.com/$WRycclIa2h__O7A4vpJHuN4jBRygLeEQMYVS3gU3JFk?via=ubuntu.com&via=matrix.org